pypeit/flatfield.py

```diff
@@ -265,7 +265,7 @@ def fit2illumflat(self, slits, frametype='illum', initial=False, flexure_shift=N
         # Loop
         for slit_idx in range(slits.nslits):
             # Skip masked
-            if slits.mask[slit_idx] != 0:
+            if (slits.mask[slit_idx] != 0) or (spat_bsplines[slit_idx].breakpoints is None):
```

Comment:
@jhennawi @profxj 
I understood a little bit better what the problem is. It happens only for morfire (or IR reduction).

The problem is when the masterflat is created. The flatfield code skips masked slits (i.e., slits with ``self.slits.mask == True``).
So in the case of a mosfire reduction, if the first science frame has a slit with failed wavelength calibration (which is computed from the science frame), the flatfield fit is skipped for that slit. But then when the reduction continues to another science frame for which the same slit has a successful wavelength calibration, the masterflat previously created will not have bspline fit for that slit and the reduction crashes.

1.  One solution could be not to skip the flatfield fit for slits with failed wavelength calibration, so that if another frame has the same slit with a successful wavelength calibration, the flatfield can be applied to that slit.
2. Or, we introduce a flatfield BPM flag for when the flatfiled fit is skipped and, if another frame has the same slit with a successful wavelength calibration, the flatfiled will still not be applied.

I think the first one could be a good solution. What do you think?

Reply:
@deborape this is not specific to MOSFIRE. It is rather specific to any situation for which we have multiple arc frames associated with science files as opposed to just one for every single file (i.e. the 'all' option). For mosfire the science is the arc, but this is also sometimes the case for non-IR instruments. Also if a user takes arcs on the sky and wants to associate them with science frames via calibration IDs, this same bug could arise. 

My understanding is that wv_calib is not currently required for flat-fielding, so I'm not sure why the flat is masked for this slit. So in that sense you proposal for 1. seems to make sense. However, we may decide to make wv_calib a dependency if we decide to do something about the vignetting issue that @marlageha and @profxj were discussing. This is not in the code yet, so I'm fine with 1).

Reply:
@jhennawi  @profxj  

I have decided to go with the option 2 and give a flat field-specific BPM flag to the slit for which the flat field is not performed because of bad wave or tilts calibration. In this way the masking will be propagated to every frames.
The bad thing about this is that if there is only one frame for which one slit has bad wave or tilts calibration all the other frames will have that slit masked. However, it make sense if these frames are combined or subtracted with each other.

I think this is the only option at the moment since the tilts image is used during the flat fielding. I made the change and tested on the dataset of the PypeIt user and resolve the issue.
I'll run the tests.

Reply:
Hi @deborape can we try to discuss via Slack. Basically I agree with your approach except for the following point: 

The dependency of flat fielding on wave_calib is not necessary and was added only to deal with the a not so common case where the user wants to modify the flat in a wavelength dependent way. There is nothing wrong with this in principle, but you can see how adding this as a required argument rather than an optional argument then leads to other undesirable side-effects. Now if a single wavelength calibration fails for a slit, it will now not be reduced for all silts as you point out. But this is very bad, since in fact wave_calib is not required for flat fielding unless the user wants to specify wavelength dependent tweaks (which is rare). This is a very good illustration of why one should think very carefully before adding dependencies on other pypeit objects as required parameters. 

My suggestion is that: 

1) wv_calib should be made optional in flat fielding. 
2) Your approach should be implemented for tilts, but not for wave_calib unless wave_calib is actually required by flat fielding. This is only if pixelflat_min_wave or pixelflat_max_wave are specified in the flat field parset. 

So basically if tilts fails for a single slit we follow your approach, but not if wavelength calib fails unless wavelength calibration was actually needed for the flat (which in the majority of cases it is not). 

The logic here is a bit subtle, so I'm happy to discuss with you and @profxj.